File: arrays/encoder/encoder.ts
import * as test from "node:test";

export class Solution {
    /**
    * @param {string} testInput
    * @return {any} Returns either boolean to validate the test cases or a custom result. 
    */
    run_solution(testInput: any) {

        const encoder = (toEncode:string[]): string=> {
            const messageLength = toEncode.map(mess => mess.length+'##').join('');
            const compressed = toEncode.join('');
            return messageLength+ compressed
        }

        const decoder = (encoded:string): string[]=> {
            const split = encoded.split("##");
            const message = split.pop();
            let prev = 0
            const decoded = split.map(len => {
                const curr = message.substring(prev, prev+Number(len))
                prev=Number(len)
                return curr
            })
            return decoded
        }

        const encoded = encoder(testInput);
        const decoded = decoder(encoded);
        console.log(decoded,testInput)

        return [testInput, decoded]
    }
}

File: arrays/encoder/encoder.py
""" Solution Module """


class Solution:
    """ Solution Class """

    def run_solution(self, test_input):
        """ Solution Method """
        # Take an input encode it and then decode it.

        def encode(to_encode: list[str]) -> str:
            len_arr = [str(len(mess))+'##' for mess in to_encode]
            prefix = ''.join(len_arr)
            stringified_message = "".join(to_encode)
            encoded = prefix+stringified_message
            return encoded

        def decode(encoded: str) -> list[str]:
            split_arr = encoded.split("##")
            encoded_message = split_arr[-1]
            len_arr = split_arr[0:len(split_arr) -1]
            print(len_arr, encoded_message)
            decoded = []
            prev_l = 0
            for l in len_arr:
                decoded.append(encoded_message[prev_l:prev_l+int(l)])
                prev_l += int(l)
            return decoded

        encoded = encode(test_input)
        decoded = decode(encoded)
        return test_input == decoded
